### PR TITLE
[REVIEW] Replace function macros with inline functions

### PR DESCRIPTION
**Pull request by kaatish**
Macros in block_utils replaced with inline device functions.

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Review by vuule — CHANGES_REQUESTED** [review]

File: cpp/src/io/utilities/block_utils.cuh [hunk 1]
```diff
@@ -19,19 +19,26 @@
...
+
+template <typename T>
+inline __device__ T shuffle_xor(T var, uint32_t delta)
+{
+  return __shfl_xor_sync(~0, var, delta);
+}
+
+inline __device__ uint32_t ballot(int pred) { return __ballot_sync(~0, pred); }
 
 #if (__CUDA_ARCH__ >= 700)
```
[the remark this answers is not in the document]
Comment (cwharris) [reply cpp/src/io/utilities/block_utils.cuh:43]: Yeah, we only support Volta and later, right?

https://docs.nvidia.com/cuda/cuda-compiler-driver-nvcc/index.html#virtual-architecture-feature-list

**Review by cwharris — APPROVED** [review]
lgtm :tada:

File: cpp/src/io/utilities/block_utils.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/io/utilities/block_utils.cuh:43]: According to the CMake its not always >= 700.

```
  # Check for embedded vs workstation architectures
  if(CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64")
    # This is being built for Linux4Tegra or SBSA ARM64
    set(GPU_ARCHS "62")
    if((CUDA_VERSION_MAJOR EQUAL 9) OR (CUDA_VERSION_MAJOR GREATER 9))
      set(GPU_ARCHS "${GPU_ARCHS};72")
    endif()
    if((CUDA_VERSION_MAJOR EQUAL 11) OR (CUDA_VERSION_MAJOR GREATER 11))
      set(GPU_ARCHS "${GPU_ARCHS};75;80")
    endif()

  else()
    # This is being built for an x86 or x86_64 architecture
    set(GPU_ARCHS "60")
    if((CUDA_VERSION_MAJOR EQUAL 9) OR (CUDA_VERSION_MAJOR GREATER 9))
      set(GPU_ARCHS "${GPU_ARCHS};70")
    endif()
    if((CUDA_VERSION_MAJOR EQUAL 10) OR (CUDA_VERSION_MAJOR GREATER 10))
      set(GPU_ARCHS "${GPU_ARCHS};75")
    endif()
    if((CUDA_VERSION_MAJOR EQUAL 11) OR (CUDA_VERSION_MAJOR GREATER 11))
      set(GPU_ARCHS "${GPU_ARCHS};80")
    endif()

  endif()
```

File: cpp/src/io/utilities/block_utils.cuh [hunk 2]
```diff
@@ -43,67 +50,67 @@ namespace io {
 template <typename T>
 inline __device__ T WarpReduceSum2(T acc)
```
[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/io/utilities/block_utils.cuh:45]: It looks like only a half warp and full warp reduction sum functions are needed. I have removed the other functions and kept those two as `WarpReduceSum` and `HalfWarpReduceSum`. Let me know if you want to keep the rest.

File: cpp/src/io/utilities/block_utils.cuh (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/io/utilities/block_utils.cuh:37]: Pascal is our minimum architecture for now.

File: cpp/src/io/orc/stripe_enc.cu [hunk 3]
```diff
@@ -402,16 +402,11 @@ static __device__ uint32_t IntegerRLE(
...
-      vmin = min(vmin, (T)shuffle_xor(vmin, 4));
-      vmin = min(vmin, (T)shuffle_xor(vmin, 8));
-      vmin = min(vmin, (T)shuffle_xor(vmin, 16));
-      vmax = max(vmax, (T)shuffle_xor(vmax, 1));
-      vmax = max(vmax, (T)shuffle_xor(vmax, 2));
-      vmax = max(vmax, (T)shuffle_xor(vmax, 4));
-      vmax = max(vmax, (T)shuffle_xor(vmax, 8));
-      vmax = max(vmax, (T)shuffle_xor(vmax, 16));
+#pragma unroll 5
+      for (int i = 1; i <= 16; i *= 2) {
```
Comment (jrhemstad) [inline cpp/src/io/orc/stripe_enc.cu:406]: This could be replaced by a `transform_reduce` with a `minmax` reduction operation.

Comment (jrhemstad) [reply cpp/src/io/orc/stripe_enc.cu:406]: Better yet, is this just a warp-level min&max reduction? Just use a CUB warp reduce.

File: cpp/src/io/orc/stripe_data.cu [hunk 4]
```diff
@@ -1206,11 +1206,11 @@ extern "C" __global__ void __launch_bounds__(NTHREADS)
...
-        skip_count += SHFL_XOR(skip_count, 1);
-        skip_count += SHFL_XOR(skip_count, 2);
-        skip_count += SHFL_XOR(skip_count, 4);
-        skip_count += SHFL_XOR(skip_count, 8);
-        skip_count += SHFL_XOR(skip_count, 16);
+        skip_count += shuffle_xor(skip_count, 1);
+        skip_count += shuffle_xor(skip_count, 2);
+        skip_count += shuffle_xor(skip_count, 4);
+        skip_count += shuffle_xor(skip_count, 8);
+        skip_count += shuffle_xor(skip_count, 16);
```
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/io/orc/stripe_data.cu]: These are just warp level reductions. These should use CUB warp reductions which are optimized for certain operations to be faster than a shuffle based reduction like this.

File: cpp/src/io/utilities/block_utils.cuh (hunk 2 above)
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/io/utilities/block_utils.cuh]: Just use CUB warp reduce?

File: cpp/src/io/orc/stripe_data.cu (hunk 4 above)
[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/io/orc/stripe_data.cu]:
```
using warp_reduce = cub::WarpReduce<int> ;
__shared__ typename WarpReduce::TempStorage temp_storage;
auto const warp_skip_count = warp_reduce(temp_storage).Sum(skip_count);
```

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/orc/stripe_data.cu]: You're referring to this https://nvlabs.github.io/cub/classcub_1_1_warp_reduce.html?
Pretty cool. if I'm reading this right, we can use it for half warp reductions too.

[the remark this answers is not in the document]
Comment (kaatish) [reply cpp/src/io/orc/stripe_data.cu]: @jrhemstad Just to clarify a half warp reduction would done like this?
`  typedef cub::WarpReduce<int, 16> WarpReduce;`

[the remark this answers is not in the document]
Comment (jrhemstad) [reply cpp/src/io/orc/stripe_data.cu]: Yeah, I'd explicitly call it a `HalfWarpReduce` though.

File: cpp/src/io/parquet/page_enc.cu [hunk 5]
```diff
@@ -1088,8 +1088,8 @@ __global__ void __launch_bounds__(128) gpuDecideCompression(EncColumnChunk *chun
         if (comp_out[comp_idx].status != 0) { atomicAdd(&error_count, 1); }
       }
     }
-    uncompressed_data_size = WarpReduceSum32(uncompressed_data_size);
-    compressed_data_size   = WarpReduceSum32(compressed_data_size);
+    uncompressed_data_size = WarpReduceSum(uncompressed_data_size);
+    compressed_data_size   = WarpReduceSum(compressed_data_size);
```
Comment (kaatish) [inline cpp/src/io/parquet/page_enc.cu:1092]: This WarpReduceSum is called inside an if branch `if (t < 32)` where `t` is `threadIdx.x`. @vuule Does this sound right? Should this not be an issue?

Comment (vuule) [reply cpp/src/io/parquet/page_enc.cu:1092]: I can't see the issue, first 32 threads will call this and they make up the warp that will contribute to the reduction. What am I missing?
cc @devavret as he's way more familiar with this code.

File: cpp/src/io/orc/stripe_data.cu [hunk 6]
```diff
@@ -1340,21 +1331,17 @@ static __device__ void DecodeRowPositions(orcdec_state_s *s, size_t first_row, i
...
       // last row
       last_row = (nz_count == s->u.rowdec.nz_count) ? row_plus1 : 0;
-      last_row = max(last_row, SHFL_XOR(last_row, 1));
-      last_row = max(last_row, SHFL_XOR(last_row, 2));
-      last_row = max(last_row, SHFL_XOR(last_row, 4));
-      last_row = max(last_row, SHFL_XOR(last_row, 8));
-      last_row = max(last_row, SHFL_XOR(last_row, 16));
+      for (int width = 1; width <= 16; width *= 2) {
+        last_row = max(last_row, shuffle_xor(last_row, width));
+      }
```
Comment (vuule) [inline cpp/src/io/orc/stripe_data.cu]: You should be able to use [this](https://nvlabs.github.io/cub/classcub_1_1_warp_reduce.html#a0dd72fc4cf7e1ecf59e8b15bd6819185) here

File: cpp/src/io/orc/stripe_data.cu [hunk 7]
```diff
@@ -1340,21 +1331,17 @@ static __device__ void DecodeRowPositions(orcdec_state_s *s, size_t first_row, i
...
       if (t < 32) {
         last_row = (t < NWARPS) ? *(volatile uint32_t *)&s->u.rowdec.last_row[t] : 0;
-        last_row = max(last_row, SHFL_XOR(last_row, 1));
-        last_row = max(last_row, SHFL_XOR(last_row, 2));
-        last_row = max(last_row, SHFL_XOR(last_row, 4));
-        last_row = max(last_row, SHFL_XOR(last_row, 8));
-        last_row = max(last_row, SHFL_XOR(last_row, 16));
+        for (int width = 1; width <= 16; width *= 2) {
+          last_row = max(last_row, shuffle_xor(last_row, width));
+        }
```
Comment (vuule) [inline cpp/src/io/orc/stripe_data.cu]: Same here, wapr_reduce + Reduce + Max should work

File: cpp/src/io/parquet/page_data.cu [hunk 8]
```diff
@@ -1269,15 +1279,17 @@ static __device__ void gpuUpdateValidityOffsetsAndRowIndices(int32_t target_inpu
...
       // for nested schemas, it's more complicated.  This warp will visit 32 incoming values,
       // however not all of them will necessarily represent a value at this nesting level. so the
       // validity bit for thread t might actually represent output value t-6. the correct position
       // for thread t's bit is cur_value_count. for cuda 11 we could use __reduce_or_sync(), but
       // until then we have to do a warp reduce.
       else {
-        warp_valid_mask = WarpReduceOr32(is_valid << thread_value_count);
+        BitWiseOR bitwise_or;
+        warp_valid_mask =
+          warp_reduce(temp_storage).Reduce(is_valid << thread_value_count, bitwise_or);
```
Comment (vuule) [inline cpp/src/io/parquet/page_data.cu:1292]:
```suggestion
        warp_valid_mask =
          warp_reduce(temp_storage).Reduce(is_valid << thread_value_count, BitWiseOR{});
```

**Review by vuule — CHANGES_REQUESTED** [review]
~~A few suggestions, mostly just inconsistent CUB use.~~
Got resolved before I posted the review. Only few nitpicks left :)

File: cpp/src/io/orc/stripe_data.cu (hunk 6 above)
Comment (kaatish) [reply cpp/src/io/orc/stripe_data.cu]: Yep. Had missed that one.

File: cpp/src/io/parquet/page_data.cu (hunk 8 above)
Comment (kaatish) [reply cpp/src/io/parquet/page_data.cu]: Fixed

**Review by vuule — APPROVED** [review]

**Comment by vuule** [comment]
rerun tests

**Comment by cwharris** [comment]
Unfortunately this PR was reverted in #6289 due to an intermittent build breakage in centos.

**Pull request merged**